RFD.md

```diff
@@ -0,0 +1,217 @@
+  - `user` - can submit any commands to the service, but can only list and see output from their own commands
+
+In an actual production system, we'd probably want all authorization tokens to have some kind of TTL and would need a way to manage tokens on the server that did not require us to rebuild the server. A more flexible role system may be needed as well.
+
+#### Data Protection
+
+The data passing through the system needs to be protected in-flight (while being transferred between the client and the server) and at-rest (command logs need to be protected if persisted on disk). Additionally, users should only be able to see output from their own commands (unless an admin-level access token is used).
+
+##### Scope limits
+
+For this exercise, we're going to assume the following constraints:
+
+* In-flight content protection will be implemented by encrypting all traffic.
+* On-disk logs protection will be limited to file permissions (0600).
+* Token-based authorization will be used to control who can see output from which commands (see the dedicated section above).
+
+In a real production scenario we may want to encrypt on-disk logs (either do it ourselves or provide the operator with guidelines on how to achieve it via `dm-crypt`, encrypted EBS, or other technologies).
+
+#### Container isolation
+
+Since the system will be used to execute arbitrary commands from the users, we need to ensure full isolation of the user command from the underlying host OS. We're going to use Linux kernel namespaces to achieve isolation along the following dimensions:
+
+* PID namespace - isolates user commands from the host OS process namespace to avoid leaking process information from the host into the container.
+
+* Network namespace - isolates container networking stack from that of the host OS.
+
+* Mount namespace - allows us to fully isolate command's filesystem from that of the host OS (and protect host OS `/proc` from being accessed by the user command to gain access to any privileged information on the host).
+
+##### Scope limits
+
+For this exercise, we're going to accept the following constraints:
+
+* Networking namespace will be used to separate the user command from the host OS network, but we're not going to build any bridges, etc to connect container network to the host.
+
+* We're going to create a new clean filesystem by extracting an Alpine Root FS image per command, which will use 5Mb of disk space per container. In a production environment we would probably rely on some kind of layered filesystem like overlayfs to save on disk resources.
+
+#### Availability and resource limits
+
+In a multi-tenant environment like the system in question, we need to ensure proper resource limits to reduce the potential impact of noisy or abusive clients on the underlying host OS.
+
+We're going to use linux Cgroups to apply the following limits on each command:
```

Comment:
Could you add the exact cgroups you plan to use below?

Reply:
Added clarifications on which cgroup controllers I was planning to use.